### PR TITLE
added instructions for deploying an applicaiton using jenkins

### PR DESCRIPTION
**Pull request by rhoads-zach**
@tradej @inoxx03 @Ladicek Let me know what you think

### CODE REVIEW

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 1]
```diff
@@ -0,0 +1,40 @@
+= Configuring Jenkins to deploy your application with the Fabric8 Maven Plugin
```
Comment (Ladicek) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Do you consider the word "plugin" to be part of the name? When I write it in full form, I always write "Fabric8 Maven plugin", but maybe I'm doing it wrong :-)

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+Similar to using maven and the Fabric8 Maven Plugin from your local host to deploy an application, you can configure Jenkins to use maven and the Fabric8 Maven Plugin to deploy an application.
```
Comment (Ladicek) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Maven is a name, capital M.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+[discrete]
+== Prerequisites
+* Access to an OpenShift cluster.
+* Jenkins installed and configured to access the same OpenShift cluster.
+* JDK and Maven installed and configured on your Jenkins server.
+* An application available in GitHub and configured to use maven and the Fabric8 Maven Plugin.
```
Comment (Ladicek) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: The prerequisites are pretty extensive. I mean, especially the 2nd one -- "Jenkins installed and configured to access the same OpenShift cluster". It's by no means clear if that expects Jenkins installed inside OpenShift or outside, but that makes a lot of difference later in the process.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc (hunk 1 above)
Comment (tradej) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Hmm, good question. I usually write _plugin_ in this context in lower case too. I've looked at their homepage, and their docs lists it all lower case (as in _fabric8 maven plugin_), so we can't really use that as a reference point either.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc (hunk 3 above)
Comment (tradej) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: I'm for clarifying point 2, but the rest seems good to me. These are the general requirements, and either we expect a specific procedure to be present earlier in the assembly that tells you how to do that specific thing, or a general knowledge how to do it if this module is presented by itself (not a big use case for the latter, we'll publish everything in assemblies primarily).

Comment (Ladicek) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Yea, right, my objection is mainly against point 2. Point 3 could use a bit of explanation as well -- you do need to do it in a traditional Jenkins setup, but if you deploy Jenkins to OpenShift, or at least let it create agents on OpenShift, you don't really have to care about installing JDK and Maven in Jenkins, you only have to make sure the Jenkins agent image has them.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc (hunk 1 above)
Comment (rhoads-zach) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: we usually say or write `fmp` or `FMP` so thats why i considered it part of the name. I don't have a strong preference if you all think it should be lower case.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+[discrete]
+== Procedure 
+. Create a new OpenShift project for your application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Do we expect the user to be able to do this on their own? If so, disregard this comment. If not, maybe a little expansion would be in order.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+. Ensure Jenkins has access to that project.
++
+For example, if you configured a service account for Jenkins, ensure that account has `edit` access to the project of your application.
+
+. Create a new link:https://wiki.jenkins.io/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-Settinguptheproject[freestyle Jenkins project] on your Jenkins server.
+.. Under _Source Code Management_, choose _Git_ and add the GitHub url of your application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: The nested list should be ushered by a colon at the end of the sentence in the first-level one (Chicago Manual of Style). I'd also prefer if the first item in this list was clicking on "New Project" or whatever the link text is, for completeness. I won't hold the PR on it, though.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+----
+clean fabric8:deploy -Popenshift -Dfabric8.namespace=MY_PROJECT
+----
++
+Ensure `MY_PROJECT` is the name of the OpenShift project for your application.
+.. Click _Save_.
+
+. Click _Build Now_ from the main page of the Jenkins project to verify your application builds and deploys to the OpenShift project for your application.
++
+You can also verify your application is deployed by opening the route in the OpenShift project of the application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: I would add "You can also verify _that_ your application is deployed..." for clarity. (low prio)

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+
+[discrete]
+== Additional resources
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Given the contents of this subsection, I'd argue it's more _Next steps_ than _Additional resources._

File: docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc [hunk 8]
```diff
@@ -0,0 +1,43 @@
+= Configuring Jenkins to deploy your {Node} application with nodeshift
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc]: Anchor ID missing.

File: docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+[discrete]
+== Procedure 
+. Create a new OpenShift project for your application.
+
+. Ensure Jenkins has access to that project.
++
+For example, if you configured a service account for Jenkins, ensure that account has `edit` access to the project of your application.
+
+. Create a new link:https://wiki.jenkins.io/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-Settinguptheproject[freestyle Jenkins project] on your Jenkins server.
+.. Under _Source Code Management_, choose _Git_ and add the GitHub url of your application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc]: Same as in `proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc`.

File: docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+.. Under _Build_, choose _Add build step_ and select `Execute Shell`.
+.. Add the following to the _Command_ area:
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+npm install -g nodeshift
+nodeshift --strictSSL=false --nodeVersion=8.x --namespace=MY_PROJECT
+----
++
+Ensure `MY_PROJECT` is the name of the OpenShift project for your application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc]: Same as in `proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc`.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+.. Under _Source Code Management_, choose _Git_ and add the GitHub url of your application.
+.. Under _Build_, choose _Add build step_ and select `Invoke top-level Maven targets`.
+.. Add the following to _Goals_:
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+clean fabric8:deploy -Popenshift -Dfabric8.namespace=MY_PROJECT
+----
++
+Ensure `MY_PROJECT` is the name of the OpenShift project for your application.
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: I find this formulation a bit unwieldy. What about: _Substitute \`MY_PROJECT\` with the name of the OpenShift project for your application._?

File: docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+.. Click _Save_.
+
+. Click _Build Now_ from the main page of the Jenkins project to verify your application builds and deploys to the OpenShift project for your application.
++
+You can also verify your application is deployed by opening the route in the OpenShift project of the application.
+
+
+[discrete]
+== Additional resources
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-to-deploy-nodejs-application-with-nodeshift.adoc]: Same as in `proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc`.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc (hunk 3 above)
Comment (rhoads-zach) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: @Ladicek re: JDK and Maven on jenkins in OpenShift, I used minishift and the Jenkins instant app to test this and I still had to install a JDK and configure JAVA_HOME for the FMP builds to work.

Comment (Ladicek) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: Not sure what's a "Jenkins instant app", but I'm pretty sure the way OpenShift sets up Jenkins for the pipelines (wihch I replicate in Rourka) doesn't need that. All builds happen in agent pods, which are spun up by the Kubernetes Jenkins plugin, using by default OpenShift-provided images, which already have Java and Maven (or Node.js) installed.

But then -- the fact that we're having this discussion shows that it's more complex than just one sentence :-)

I don't necessarily think RHOAR docs must explain all this (or do they?). But it's complex.

File: docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc (hunk 4 above)
Comment (rhoads-zach) [reply docs/topics/proc_configuring-jenkins-to-deploy-an-application-with-fmp.adoc]: I'm assuming the user can do this, but i'll expand this with sub-steps

**Comment by rhoads-zach** [comment]
@tradej @inoxx03 @Ladicek addressed itemized feedback and adjusted title to make content more specific to a Jenkins Freestyle project.

Let me know what you think

File: docs/topics/proc_configuring-jenkins-freestyle-project-to-deploy-an-application-with-fmp.adoc [hunk 13]
```diff
@@ -0,0 +1,46 @@
+[id='configuring-jenkins-freestyle-project-to-deploy-your-application-fmp-{context}']
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-freestyle-project-to-deploy-an-application-with-fmp.adoc]: Please do not deviate from the title, use `configuring-a-jenkins-freestyle-project-to-deploy-your-application-with-the-fabric8-maven-plugin_{context}` as the ID.

File: docs/topics/proc_configuring-jenkins-freestyle-project-to-deploy-nodejs-application-with-nodeshift.adoc [hunk 14]
```diff
@@ -0,0 +1,49 @@
+[id='configuring-jenkins-freestyle-project-to-deploy-your-application-nodeshift-{context}']
```
Comment (tradej) [inline docs/topics/proc_configuring-jenkins-freestyle-project-to-deploy-nodejs-application-with-nodeshift.adoc]: Please do not deviate from the title, use `configuring-a-jenkins-freestyle-project-to-deploy-your-node-application-with-nodeshift_{context}` as the ID.

**Comment by rhoads-zach** [comment]
thank you @tradej for the feedback. Fixed the anchors.

@Ladicek What are your thoughts about us merging this content and adding more content later with other jenkins options (e.g. pipelines)? Or do you feel the current state is the wrong direction?

**Comment by Ladicek** [comment]
I'm totally fine with merging this, it's a good start. Maybe it should at least mention that the Jenkins job could, before deploying, execute tests. I was also thinking about adding a screenshot, but I know that every screenshot significantly increases maintenance burden, so I'll leave that idea to you.

**Comment by rhoads-zach** [comment]
thank you @Ladicek added a next step about adding test build step before deploying

**Comment by Ladicek** [comment]
LGTM, with two questions:

1. Is it "nodeshift" or "Nodeshift"? I honestly don't know, I've always used "Nodeshift", but looking at READMEs etc., it seems "nodeshift" is more common...
2. Should we explicitly mention Red Hat productized base images? For FMP, one can specify it in the `pom.xml` (that's what we do with the boosters), but with nodeshift, I think (not sure) that the only way is to specify the RHOAR Node.js image when running nodeshift.

**Comment by tradej** [comment]
Re 1: The homepage says _nodeshift,_ so I suggest we roll with that.

**Comment by rhoads-zach** [comment]
thank you @tradej @Ladicek 

agree with @tradej about `nodeshift`

I've amended the PR to include a callout the base image in the java-based runtimes.

**Pull request merged**